### PR TITLE
Breaking change: MouseTrackerAnnotation requires parameter "key"

### PR DESCRIPTION
**Pull request by dkwingsmt**
Breaking change documentation for PR https://github.com/flutter/flutter/pull/46034.
Design doc: https://flutter.dev/go/key-based-mouse-tracker-annotation

### CODE REVIEW

File: src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+dispatches callbacks accordingly. `MouseTracker` also tracks which
+annotations are on the screen by requring the annotation's owners to attach
+or detach the annotation promptly.
+
+However, since the `MouseTrackerAnnotation` is a constant class, its properties
+can't be changed without replacing the annotation as a whole, which triggers
+the exit and enter callback if it's hovered by a mouse pointer.
+
+This issue is a blocker to the plan of supporting mouse cursor, since it’s
+very likely that the cursor assigned to a region needs to be changed.
```
Comment (Hixie) [inline src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md]: We should write this from the point of view of having already made the change. There used to be a problem, the problem was bla; this change resolves that issue.

File: src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+
+## Description of change
+
+Class `MouseTrackerAnnotation` now requires a new parameter `key`:
```
Comment (Hixie) [inline src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md]: The `MouseTrackerAnnotation` class...

File: src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
++  final LocalKey key;
+
+   ...
+ }
+```
+
+Three of class `MouseTracker`'s methods, `attachAnnotation`,
+`detachAnnotation`, and `isAnnotationAttached`, instead of taking a
+single parameter of `MouseTrackerAnnotation`, will now take a
+single parameter `LocalKey`:
```
Comment (Hixie) [inline src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md]: here it gives the type of the parameter but earlier it gave the name of the parameter. we should be consistent.

File: src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+-  void detachAnnotation(MouseTrackerAnnotation annotation);
++  void detachAnnotation(LocalKey annotationKey);
+ }
+```
+
+## Migration guide
+
+There are 2 ways of migrating affected code. The first way requires little
+work, while the second is much simpler when the annotation's properties
```
Comment (Hixie) [inline src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md]: "requires little work" means it's easy
"requires a little work" means it's not easy
I think this was supposed to be the second?

File: src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+     super.attach(owner);
+-    RendererBinding.instance.mouseTracker.attachAnnotation(_hoverAnnotation);
++    RendererBinding.instance.mouseTracker.attachAnnotation(_hoverAnnotation.key);
+   }
+
+   @override
+   void detach() {
+-    RendererBinding.instance.mouseTracker.detachAnnotation(_hoverAnnotation);
++    RendererBinding.instance.mouseTracker.detachAnnotation(_hoverAnnotation.key);
+     super.detach();
```
Comment (Hixie) [inline src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md]: why does the render object do this instead of the layer tree? seems weird that you have to independently tell the layer tree and the mouse tracker about the annotations.

File: src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md (hunk 1 above)
Comment (dkwingsmt) [reply src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md]: Fixed.

File: src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md (hunk 2 above)
Comment (dkwingsmt) [reply src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md]: Fixed.

File: src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md (hunk 3 above)
Comment (dkwingsmt) [reply src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md]: Fixed.

File: src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md (hunk 4 above)
Comment (dkwingsmt) [reply src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md]: Yes. I used "requires less work" and made the difference of the 2 methods clearer.

File: src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md (hunk 5 above)
Comment (dkwingsmt) [reply src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md]: There are multiple reasons:
- Annotation attachment is used to indicate widget presence. Generally, an annotation should be detached if the owner widget is unmounted, for the reason described in https://github.com/flutter/flutter/pull/44631 "The change to onExit". Therefore I think it's better to handle it in the render object (or should it be the element?).
- Historical reason. Attaching and detaching annotations used to have immediate effects (triggering callbacks) and we had to do it during Element.activate and deactivate. This was no longer needed, so now we only need to do it in attach and detach, but we haven't fully moved it to layers yet.
- Future plan. We're planning to experiment a change that split annotations from the layer tree to fix issues such as https://github.com/flutter/flutter/issues/41194. Therefore we'd like to keep it separated from the layer tree (at least for now).

Comment (Hixie) [reply src/docs/release/breaking-changes/mousetrackerannotation-requires-parameter-key.md]: It'll still have to be in the layer tree though right? Just done differently than today (many nodes dealing with one layer rather than one layer per node).

This just seems like an unnecessarily awkward API right now.

Anyway, the point of my questions is that you should explain this in the breaking change document. Especially if we're planning on changing this again, we should warn people that we're not done changing this API.

**Comment by dkwingsmt** [comment]
Closing this because I found out that it's probably neither necessary nor sufficient. I've also closed the corresponding Flutter PR in favor of https://github.com/flutter/flutter/pull/48453.
